samcli/commands/_utils/template.py

```diff
@@ -161,6 +156,31 @@ def _update_relative_paths(template_dict, original_root, new_root):
     return template_dict
 
 
+def _transform_dict(template_dict, path, transformation, context=None):
+    def visit(sub_dict, sub_path):
+        if not sub_path:
+            return sub_dict
+
+        next_prop = sub_path[0]
+        if not next_prop in sub_dict:
+            return sub_dict
+
+        value = sub_dict[next_prop]
+        if len(sub_path) == 1:
+            sub_dict[next_prop] = transformation(template_dict, path, next_prop, value, context)
+        else:
+            if isinstance(value, list):
+                sub_dict[next_prop] = [visit(v, sub_path[1:]) for v in value]
+            elif isinstance(value, dict):
+                sub_dict[next_prop] = visit(value, sub_path[1:])
+            else:
+                raise Exception("Unsupported path '%s' or dict structure '%s'" % path, template_dict)
```

Comment:
`Exception` is too generic, you can maybe create a `DictTransformException` with `UsedException` as base class